### DISCUSSION TITLE
Suggestions for performance improvement (Best practices) when using Xarray with Dask

### DISCUSSION THREAD

**Discussion by ricardobarroslourenco** [pydata/xarray#7632] [Office Hours]
On a workflow for Remote Sensing (please refer to the [gist](https://gist.github.com/ricardobarroslourenco/922f8693f3eeeaf9fa11b49c6da19947)) I am currently able to load and do the processing of my data (originally ENVI raster files transformed into netcdf with `rioxarray`), and the results are quite good (initially, I used to process in R, facing some memory limitations and issues with data structures, and Dask+Xarray solved this end of the problem).

However, in this mentioned notebook, I am facing performance issues, especially when getting to the end of the workflow. Saving the Dataset is taking hours; however, the CPU utilization is somehow low (average in the time-series of 30-40%), and the write bandwidth is almost unused, around 2Mbps in the Dask monitoring (I am running in an Apptainer container (derived from this [Docker image](https://hub.docker.com/r/ricardobarroslourenco/mambaforge_custom_geo_cpuonly) )  in an HPC node with 128GB of RAM and saving it in a `scratch` partition which has lustre filesystem (and Gigabit speeds).

Any suggestions on how can I start improving my workflow?

**EDIT 1**: 
![Screen Shot 2023-03-15 at 14 01 09](https://user-images.githubusercontent.com/7102375/225407037-3b7803ef-636a-4d6c-91d2-e1ad066349d4.png)

A compute time per task screenshot. How can I find the functions that trigged such tasks? I was wondering if I have a margin to improve things here (basically on the four large ones, which do stacking,  and mapping on store)

**EDIT 2**: The source data can be found [here](https://drive.google.com/drive/folders/162KvwfOzRFmx7G7IPN4zYeaSJQW7HYXs?usp=share_link).

**EDIT 3**: It seems that some issues are happening with the HDF5 library... =/

![Screen Shot 2023-03-15 at 15 26 12](https://user-images.githubusercontent.com/7102375/225421640-556ac9bc-8ebb-4f04-87d7-a2b15754209b.png) 

**EDIT 4**: Adding [new version of the notebook](https://gist.github.com/ricardobarroslourenco/67f5f6cdd06520832e1429407b73a8d1), with better usage of `apply_ufunc`,  but the process of exporting to netCDF is still really slow...

**Comment by dcherian** [comment]
What is the `stack`  piece? We'll need to look at your code and input chunk sizes.

We could have a longer discussion at the [office hours on Friday](https://github.com/pydata/xarray/discussions/7564)

**Reply by ricardobarroslourenco** [reply]
Great! I will sign up then.

Meanwhile, the notebook is available here in this [gist](https://gist.github.com/ricardobarroslourenco/922f8693f3eeeaf9fa11b49c6da19947). I noticed a weird thing. When writing to disk (via a `Dataset.to_netcdf` call), the file is created, and within the first 10-15 minutes, it gets to 4.5 GB and then stops growing (Currently 1h05m elapsed). The cluster memory usage is around 7 GB.

Do you think that perhaps a dependency (for ex.: NetCDF library) may be causing the issue? The progress panel on Dask shows that many tasks are queued, but no actual file writing is happening.

**Reply by ricardobarroslourenco** [reply]
I have just registered for office hours and included the source data in the main post here.

**Reply by ricardobarroslourenco** [reply]
Just included a new revised version of the notebook, with the recent improvements we discussed.

**Reply by ricardobarroslourenco** [reply]
@dcherian just putting a note here on what we discussed today at office hours (it might be helpful to others):

- Force `.compute()` before calling `.to_netcdf()` # In my case, I think this is already in place
- Use Zarr instead of netCDF
   - Provides better parallelization
   - And netCDF format will likely adopt Zarr structure soon: https://docs.unidata.ucar.edu/nug/current/nczarr_head.html

I think I have not forgotten anything. I will try to apply these changes and see how it goes (if successful, I will post back here). Thanks again!

**Reply by dcherian** [reply]
And using flox for multiple variable groupby is probably a lot more efficient (no stacking/unstacking),

**Comment by kmuehlbauer** [comment]
@ricardobarroslourenco Regarding the HDF warnings, this might be related https://github.com/pydata/xarray/issues/7549

**Reply by ricardobarroslourenco** [reply]
Thanks for the reference and guidance @kmuehlbauer
